### PR TITLE
vk: Add support for VK_EXT_custom_border_color

### PR DESCRIPTION
**Pull request by kd-11**
Turns out some games are really using some unusual border colors (e.g transparent white)
Previous fallback will still exist in case of GPUs without support for this.
Also includes a refactor of how we handle optional features, so that a migration up is easier if we decide to raise our base vulkan version to 1.1 or higher.

Fixes https://github.com/RPCS3/rpcs3/issues/12451

### CODE REVIEW

File: rpcs3/Emu/RSX/VK/vkutils/sync.cpp [hunk 1]
```diff
@@ -286,7 +286,7 @@ namespace vk
 		vk::gpu_debug_marker::insert(
 			*m_device,
 			*m_cb,
-			fmt::format("0x%x: Enter %s", m_tag, m_message)
+			fmt::format("0x%llx: Enter %s", m_tag, m_message)
```
Comment (elad335) [inline rpcs3/Emu/RSX/VK/vkutils/sync.cpp:289]: fmt::format doesn't care about these primitive prefixes iirc,

File: rpcs3/Emu/RSX/VK/vkutils/device.cpp [hunk 2]
```diff
@@ -73,20 +74,28 @@ namespace vk
...
 
-			shader_types_support.allow_float64 = !!features2.features.shaderFloat64;
-			shader_types_support.allow_float16 = !!shader_support_info.shaderFloat16;
-			shader_types_support.allow_int8    = !!shader_support_info.shaderInt8;
-			framebuffer_loops_support          = !!fbo_loops_info.attachmentFeedbackLoopLayout;
-			barycoords_support                 = !!shader_barycentric_info.fragmentShaderBarycentric;
-			features                           = features2.features;
+			shader_types_support.allow_float64   = !!features2.features.shaderFloat64;
+			shader_types_support.allow_float16   = !!shader_support_info.shaderFloat16;
+			shader_types_support.allow_int8      = !!shader_support_info.shaderInt8;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/device.cpp:90]: these 3 lines seem to be changed for no reason, at least they could be aligned better

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp [hunk 3]
```diff
@@ -4,9 +4,46 @@
 
 namespace vk
 {
+	extern VkBorderColor get_border_color(u32);
+
+	static VkBorderColor get_closest_border_color_enum(const color4f& color4)
+	{
+		if ((color4.r + color4.g + color4.b) > 1.35f)
+		{
+			//If color elements are brighter than roughly 0.5 average, use white border
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:13]:
```suggestion
			// If color elements are brighter than roughly 0.5 average, use white border
```

als0, why 0.5 ? it's 1.35 above

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp [hunk 4]
```diff
@@ -4,9 +4,46 @@
...
+
+	border_color_t::border_color_t(u32 encoded_color)
+		: storage_key(0)
+	{
+		value = vk::get_border_color(encoded_color);
+
+		if (value != VK_BORDER_COLOR_FLOAT_CUSTOM_EXT)
+		{
+			// Nothing to do
+			return;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:31]: color_value is uninitialized in this case (unless it has a proper default constructor)

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp [hunk 5]
```diff
@@ -25,7 +62,20 @@ namespace vk
 		info.minFilter = min_filter;
 		info.mipmapMode = mipmap_mode;
 		info.compareOp = depth_compare_mode;
-		info.borderColor = border_color;
+		info.borderColor = border_color.value;
+
+		VkSamplerCustomBorderColorCreateInfoEXT custom_color_info;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:67]: doesn't this warn about initialization?

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp [hunk 6]
```diff
@@ -52,4 +102,114 @@ namespace vk
...
+				use_border_encoding = true;
+			}
+		}
+
+		key.base_key = u16(clamp_u) | u64(clamp_v) << 3 | u64(clamp_w) << 6;
+		key.base_key |= u64(unnormalized_coordinates) << 9;            // 1 bit
+		key.base_key |= u64(min_filter) << 10 | u64(mag_filter) << 11; // 1 bit each
+		key.base_key |= u64(mipmap_mode) << 12;   // 1 bit
+
+		if (!use_border_encoding)
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:131]: can be reversed

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp [hunk 7]
```diff
@@ -52,4 +102,114 @@ namespace vk
...
+	cached_sampler_object_t* sampler_pool_t::find(const sampler_pool_key_t& key) const
+	{
+		if (!key.border_color_key) [[ likely ]]
+		{
+			const auto found = m_generic_sampler_pool.find(key.base_key);
+			return found == m_generic_sampler_pool.end() ? nullptr : found->second.get();
+		}
+
+		const auto block = m_custom_color_sampler_pool.equal_range(key.base_key);
+		for (auto It = block.first; It != block.second; ++It)
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:166]: why uppercase?

**Comment by Darkhost1999** [comment]
Infamous master
image
Infamous PR
image
No more purple border 😃

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp (hunk 7 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:166]: I paid for the whole keyboard, I'll use the whole keyboard XD

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp (hunk 3 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:13]: It's about 0.45, though the target is 0.5 with a factor. I must have changed it to fix some game many years ago and forgot to update the comment.

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp (hunk 4 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:31]: It's a color4f object which default initializes to black.

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp (hunk 5 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:67]: Nope. The compiler is smart enough to see it is initialized before use.

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp (hunk 6 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:131]: The default is to not use border encoding. It is more readable this way.

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp (hunk 7 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/sampler.cpp]: Fixed.

File: rpcs3/Emu/RSX/VK/vkutils/device.cpp (hunk 2 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/device.cpp]: Fixed.

File: rpcs3/Emu/RSX/VK/vkutils/sampler.cpp (hunk 5 above)
Comment (Megamouse) [reply rpcs3/Emu/RSX/VK/vkutils/sampler.cpp:67]: "The compiler". Which one xD

**Pull request merged**